packages/cli/src/test/integration/integration_test.ts

```diff
@@ -117,14 +133,70 @@ suite('integration tests', function() {
       await runCommand(binPath, ['install'], {cwd: dir});
-      // await runCommand(
-      //   binPath, ['lint', '--rules=polymer-3'],
-      //   {cwd: dir})
-      // await runCommand(binPath, ['test'], {cwd: dir})
-      await runCommand(binPath, ['build'], {cwd: dir});
+      await Promise.all([
+        // Does not lint clean at the moment.
+        // TODO: https://github.com/Polymer/tools/issues/274
+        // runCommand(binPath, ['lint', '--rules=polymer-3'], {cwd: dir}),
+        runCommand(binPath, ['build'], {cwd: dir}),
+      ]);
+      async function assertTrueInPage(
+          page: puppeteer.Page, expression: string) {
+        assert(
+            await page.evaluate(expression),
+            `Expected \`${expression}\` to evalutate to true in the browser.`);
+      }
+      async function assertThatShopWorks(dirToServe: string) {
+        const startResult = await startServers({root: dirToServe});
+        if (startResult.kind === 'MultipleServers') {
+          for (const server of startResult.servers) {
+            server.server.close();
+          }
+          throw new Error(`Unexpected startResult`);
+        }
+        disposables.push(() => {
+          startResult.server.close();
+        });
+        const address = startResult.server.address();
+        const baseUrl = `http://${address.address}:${address.port}`;
+        const page = await (await browserPromise).newPage();
+        disposables.push(() => page.close());
+        await page.goto(`${baseUrl}/`);
+        assert.deepEqual(`${baseUrl}/`, page.url());
+        // For the moment we don't type check in-browser expressions.
+        // tslint:disable-next-line: no-any
+        type Window = any;
+        await page.waitFor(function(this: Window) {
+          return this.document.querySelector('shop-app').shadowRoot != null;
+        });
+        // The app is defined, but the cart isn't.
+        await assertTrueInPage(
+            page, `customElements.get('shop-app') !== undefined`);
+        await assertTrueInPage(
+            page, `customElements.get('shop-cart') === undefined`);
+        // Click the shopping cart button.
+        await page.evaluate(`
+            document.querySelector('shop-app').shadowRoot
+                .querySelector('a[href="/cart"]')
+                .click()`);
+        // Url changes immediately
+        assert.deepEqual(`${baseUrl}/cart`, page.url());
+        // We'll lazy load the code for shop-cart. We'll know that it worked
+        // when the element is registered. If this resolves, it loaded
+        // successfully!
+        await page.waitFor(function(this: Window) {
+          return this.customElements.get('shop-cart') !== undefined;
+        });
+      }
+      const config =
+          ProjectConfig.loadConfigFromFile(path.join(dir, 'polymer.json'));
+      if (config == null) {
+        throw new Error('Failed to load shop\'s polymer.json');
+      }
+      await Promise.all(config.builds.map(
```

Comment:
Looks like it's not going to be obvious which build was running when any assertion failed, is that right? Would be nice to define a different test for each build, although I guess then it won't run in parallel. Update all the assert calls to mention the build or something?

Reply:
Yeah the most important thing is really just avoiding repeating the build action, but it's a bit faster to run the browser tests in parallel